src/org/traccar/BaseProtocolDecoder.java

```diff
@@ -31,6 +31,22 @@
 
     private final Protocol protocol;
 
+    public long addUnknownDevice(String uniqueId, SocketAddress remoteAddress) {
+        long deviceId = 0;
+        Device newDevice = new Device();
+        newDevice.setName(uniqueId);
+        newDevice.setUniqueId(uniqueId);
+
+        try {
+            Context.getDeviceManager().addDevice(newDevice);
+            Log.info("Nice meeting new device " + uniqueId + ".");
+        } catch (Exception e) {
+            Log.warning(e);
+        }
+
+        return deviceId;
```

Comment:
Sorry for interrupting.
Am I missing something or it will always return 0? I guess you need set `deviceId` after you added a device.